crates/rethnet_rpc_server/src/lib.rs

```diff
@@ -1 +1,200 @@
+                                data: Some(serde_json::to_value(version).unwrap()),
+                            }}
+                        }))
+                    },
+                    RpcRequest { version: _, id, method } => {
+                        match method {
+                            MethodInvocation::GetBalance(address, _block_spec) => {
+                                Json(serde_json::json!(jsonrpc::Response {
+                                    jsonrpc: jsonrpc::Version::V2_0,
+                                    id,
+                                    data: match (*(rethnet_state.lock().await)).basic(address) {
+                                        Ok(Some(account_info)) => {
+                                            jsonrpc::ResponseData::<U256>::Success { result: account_info.balance }
+                                        },
+                                        Ok(None) => {
+                                            jsonrpc::ResponseData::<U256>::Error { error: jsonrpc::Error {
+                                                code: 0, message: String::from("No such account"), data: None,
+                                            }}
+                                        }
+                                        Err(e) => {
+                                            jsonrpc::ResponseData::<U256>::Error { error: jsonrpc::Error {
+                                                code: 0, message: e.to_string(), data: None,
+                                            }}
+                                        }
+                                    }
+                                }))
+                            }
+                            _ => { // TODO: after adding all the methods here, eliminate this
+                                   // catch-all match arm.
+                                Json(serde_json::json!(jsonrpc::Response {
+                                    jsonrpc: jsonrpc::Version::V2_0,
+                                    id,
+                                    data: jsonrpc::ResponseData::<serde_json::Value>::Error { error: jsonrpc::Error {
+                                        code: 0,
+                                        message: String::from("unsupported JSON-RPC method"),
+                                        data: Some(serde_json::to_value(method).unwrap()),
+                                    }}
+                                }))
+                            }
+                        }
+                    }
+                }
+            }
+        )
+    )
+    .with_state(state)
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use axum::{body::Body, http::Request};
+    use rethnet_eth::{
+        remote::{jsonrpc, BlockSpec, MethodInvocation},
+        Address,
+    };
+    use rethnet_evm::KECCAK_EMPTY;
+    use revm::primitives::state::AccountInfo;
+    use std::net::{SocketAddr, TcpListener};
+
+    fn start_server() -> SocketAddr {
```

Comment:
I'd move part of this function into the lib as `async fn run(addr: SocketAddr, genesis_accounts: HashMap<Address, AccountInfo>)`, which runs the `axum::Server`. We'll need that once we wrap this in the `rethnet` CLI too.

Reply:
addressed in https://github.com/NomicFoundation/hardhat/pull/3998/commits/07c267afd1162739fbe13cc16ce74af6c5e56238